### PR TITLE
Internalizing public types

### PR DESCRIPTION
**Pull request by SeanFeldman**
## Who's affected
- Anyone extending and customizing Azure Service Bus transport version 7 and above
## Symptoms

Version 7.0 of the ASB transport has exposed internal types that shouldn't have been exposed.
## Original issue

<details><summary>

Details</summary>

The following types have been exposed and broken down into several categories for a review.

**Types that should be exposed**

✔️ IIndividualizationStrategy
✔️ IHandleOversizedBrokeredMessages
✔️ ISanitizationStrategy
✔️ ICompositionStrategy
✔️ INamespacePartitioningStrategy

**Types that should NOT have been exposes at all**

❌ ITopology
❌ IStoppableTopology
❌ ITransportPartsContainer
❌ IResolveTransportParts  
❌ IRegisterTransportParts
❌ INotifyIncomingMessages
❌ IClientSideSubscriptionFilter
❌ INamespaceManager ==|> NamespaceManagerAdapter

**Types that require intimate knowledge of topologies and/or transport**

❗️ ICreateAzureServiceBusSubscriptions ==|> AzureServiceBusSubscriptionCreator + AzureServiceBusSubscriptionCreatorV6 / AzureServiceBusForwardingSubscriptionCreator (creator per topology)
❗️  ITopologySectionManager ==|> EndpointOrientedTopologySectionManager / ForwardingTopologySectionManager
❗️ IOperateTopology ==|> TopologyOperator
❗️ IBatcher ==|> Batcher / Batch / BatchedOperation
❗️ IRouteOutgoingBatches ==|> DefaultOutgoingBatchRouter / ReceiveContext
❗️ IManageMessageSenderLifeCycle ==|> MessageSenderLifeCycleManager
❗️ ICreateMessageSenders ==|> MessageSenderCreator
❗️ IManageMessagingFactoryLifeCycle ==|> MessagingFactoryLifeCycleManager
❗️ ICreateMessagingFactories ==|> MessagingFactoryCreator
❗️ ICreateMessageReceivers ==|> MessageReceiverCreator
❗️ IManageMessageReceiverLifeCycle ==|> MessageReceiverLifeCycleManager
❗️ IConvertBrokeredMessagesToIncomingMessages ==|> DefaultBrokeredMessagesToIncomingMessagesConverter is the default. Custom implementation not possible since it would require ICanMapConnectionStringToNamespaceAlias which is an internal contract
❗️ IConvertOutgoingMessagesToBrokeredMessages ==|> DefaultBatchedOperationsToBrokeredMessagesConverter

**Types in question**

❓ ICreateTopology ==|> TopologyCreator
❓ IManageNamespaceManagerLifeCycle ==|> NamespaceManagerLifeCycleManager
❓ ICreateNamespaceManagers ==|> NamespaceManagerCreator
❓ ICreateAzureServiceBusQueues ==|> AzureServiceBusQueueCreator
❓ ICreateAzureServiceBusTopics ==|> AzureServiceBusTopicCreator
❓ IClientEntity ==|> IMessagingFactory / IMessageReceiver / IMessageSender 
❓ IMessagingFactory ==|> MessagingFactoryAdapter
❓ IMessageReceiver ==|> MessageReceiverAdapter
❓ IMessageSender ==|> MessageSenderAdapter
❓ IBrokerSideSubscriptionFilter ==|> only exposed because of the public SubscriptionInfo, which is exposed via TopologySection revealed by ITopologySectionManager
---
#### `ITransportPartsContainer`, `IRegisterTransportParts`, `IResolveTransportParts`
- Default container implementation (`TransportPartsContainer`) was not intended for replacement. 
- Access to the default container is revealed via dependency resolution, which can lead to undesired results.
#### `INotifyIncomingMessages`

`TopologyOperator` creates `INotifyIncomingMessages`, but it's [coupled to `MessageReceiverNotifier` implementation](https://github.com/Particular/NServiceBus.AzureServiceBus/blob/01b0b804c69d5d085e78d092d1b78f876fa84914/src/Transport/Topology/TopologyOperator.cs#L117) only. Having `INotifyIncomingMessages` exposed adds no value.
#### `IStoppableTopology`

To be merged with `ITopology` in 8.0.0
#### `INamespaceManager`

Default implementation provided by `NamespaceManagerAdapter` is an abstraction on top of native `NamespaceManager` and additional functionality required to determine if management is allowed for a given namespace.
#### `IClientSideSubscriptionFilter`

Exposed via `SubscriptionInfo.ClientSideFilter` that is never used

 

**Update**: agreed to internalize everything marked in red and review need to expose things based on case-by-case need.

</details>

### CODE REVIEW

**Comment by SeanFeldman** [comment]
I've obsoleted a few types that should not be exposed (too destructive or never used).
Updated description with all the public contracts (except models) for discussion and review.

/cc @yvesgoeleven @danielmarbach

**Comment by SeanFeldman** [comment]
Somewhat related issue #390 - Internal contracts with a single implementation only

**Comment by SeanFeldman** [comment]
@Particular/azure-service-bus-maintainers please review. 

There were contracts and models that needed to be obsoleted for internalization. For tests, since we have "treat all warnings as errors", had to use `#pragma warning disable 618`.

File: src/Tests/Creation/When_creating_forwarding_subscription.cs [hunk 1]
```diff
@@ -57,7 +58,7 @@ public void TopicCleanUp()
             const string subscriptionName = "endpoint0";
             //make sure there is no leftover from previous test
             var namespaceManager = new NamespaceManagerAdapter(NamespaceManager.CreateFromConnectionString(AzureServiceBusConnectionString.Value));
-            await namespaceManager.DeleteSubscriptionAsync(topicPath, subscriptionName);
+            await namespaceManager.DeleteSubscription(new SubscriptionDescription(topicPath, subscriptionName));
```
Comment (danielmarbach) [inline src/Tests/Creation/When_creating_forwarding_subscription.cs]: Why is the delete subscription contained in this PR? I suppose those changes are unrelated to this PR?

File: src/Tests/Creation/When_creating_subscription.cs [hunk 2]
```diff
@@ -45,7 +46,7 @@ public void TopicCleanUp()
             const string subscriptionName = "mysubscription1";
             //make sure there is no leftover from previous test
             var namespaceManager = new NamespaceManagerAdapter(NamespaceManager.CreateFromConnectionString(AzureServiceBusConnectionString.Value));
-            await namespaceManager.DeleteSubscriptionAsync(topicPath, subscriptionName);
+            await namespaceManager.DeleteSubscription(new SubscriptionDescription(topicPath, subscriptionName));
```
Comment (danielmarbach) [inline src/Tests/Creation/When_creating_subscription.cs]: Why is the delete subscription contained in this PR? I suppose those changes are unrelated to this PR?

File: src/Transport/Connectivity/ICreateMessageReceivers.cs [hunk 3]
```diff
@@ -2,6 +2,7 @@ namespace NServiceBus.Transport.AzureServiceBus
 {
     using System.Threading.Tasks;
 
+    [ObsoleteEx(Message = "Internal contract that shouldn't be exposed.", TreatAsErrorFromVersion = "8.0", RemoveInVersion = "9.0")]
```
Comment (danielmarbach) [inline src/Transport/Connectivity/ICreateMessageReceivers.cs]: How about moving the message to a constant? Then we can change it in one place?

File: src/Transport/Connectivity/NamespaceManagerAdapter.cs [hunk 4]
```diff
@@ -113,9 +114,15 @@ public Task<TopicDescription> GetTopic(string path)
...
         }
 
+        // TODO: delete this once NamespaceManagerAdapter is internalized
         public Task DeleteSubscriptionAsync(string topicPath, string subscriptionName)
         {
-            return manager.DeleteSubscriptionAsync(topicPath, subscriptionName);
+            return DeleteSubscription(new SubscriptionDescription(topicPath, subscriptionName));
+        }
+
+        public Task DeleteSubscription(SubscriptionDescription subscriptionDescription)
```
Comment (danielmarbach) [inline src/Transport/Connectivity/NamespaceManagerAdapter.cs]: Why is the delete subscription contained in this PR? I suppose those changes are unrelated to this PR?

File: src/Transport/Creation/AzureServiceBusForwardingSubscriptionCreator.cs [hunk 5]
```diff
@@ -135,6 +135,46 @@ public AzureServiceBusForwardingSubscriptionCreator(ReadOnlySettings settings)
 
         }
 
+        public async Task DeleteSubscription(string topicPath, string subscriptionName, SubscriptionMetadata metadata, string sqlFilter, INamespaceManager namespaceManager, string forwardTo)
```
Comment (danielmarbach) [inline src/Transport/Creation/AzureServiceBusForwardingSubscriptionCreator.cs]: Why is the delete subscription contained in this PR? I suppose those changes are unrelated to this PR?

File: src/Transport/Topology/ICreateTopology.cs [hunk 6]
```diff
@@ -7,8 +7,15 @@ namespace NServiceBus.Transport.AzureServiceBus
...
     // note there is some creation logic elsewhere already, those calls should be removed and centralized here
 
+    [ObsoleteEx(Message = "Internal contract that shouldn't be exposed.", TreatAsErrorFromVersion = "8.0", RemoveInVersion = "9.0")]
     public interface ICreateTopology
     {
         Task Create(TopologySection topology);
     }
+
+    // TODO: Move into internalized ICreateTopology in v8
+    interface ICreateTopologyAbleToDeleteSubscriptions
```
Comment (danielmarbach) [inline src/Transport/Topology/ICreateTopology.cs]: Why is the delete subscription contained in this PR? I suppose those changes are unrelated to this PR?

File: src/Transport/Topology/TopologyCreator.cs [hunk 7]
```diff
@@ -51,5 +51,36 @@ public TopologyCreator(ITransportPartsContainer container, IManageNamespaceManag
                 }
             }
         }
+
+        public async Task TearDownSubscription(TopologySection topologySection)
```
Comment (danielmarbach) [inline src/Transport/Topology/TopologyCreator.cs]: Why is the delete subscription contained in this PR? I suppose those changes are unrelated to this PR?

**Comment by SeanFeldman** [comment]
@Particular/azure-service-bus-maintainers please review and double check that we're covering all the public types that need to be internalized.

**Comment by SeanFeldman** [comment]
@Particular/azure-service-bus-maintainers final review call :)
